### PR TITLE
PIM-6353: Integration tests - Indexing/Researching products and models in ES

### PR DESCRIPTION
**Pull request by SamirBoulil**
**Description (for Contributor and Core Developer)**

This PR adds only the ES integration tests regarding the indexing of products, product variants and product models in ES in order to validate that it is possible to:
- Effectively search for them in the Datagrid
- Effectively export products via the export builder.

**Indexing of products:**
- The products/product variants and product models are indexed in the same index (for now) and index type.
- The values indexed for each documents (products & product models) refer to the attributes they own `union` all the product values of their parents.
- There are new properties added in the ES mapping: 
   - "product_type" and help us differenciate between products, models. The possible value may change in the future, but for now are `'PimCatalogProductVariant', 'PimCatalogProduct', 'PimCatalogSubProductModel', 'PimCatalogRootProductModel'`
   - "own_attributes": Help us know which property is really owned by the document and not inherited by a model.

**Dataset**:
We recreated @Delphineray 's dataset in the https://github.com/akeneo/pim-community-dev/pull/6451/files#diff-5592369bc813e82c56e826b8f9591a80 . 

It shows how each products and product models is indexed in ES using PHP arrays.

**Searching for products**:
While executing an export job, we need to make sure the PQB only returns products instances (simple products and product variants).

In the`https://github.com/akeneo/pim-community-dev/pull/6451/files#diff-d1794244d0337590cd3afaa5b103f84e` integration tests you can see how the ES request look like.
In the end, they are regular ES requests we use to write with an additional condition on the whole new "product_type" property (e.g: we only want products and product_variants).

**Searching for products and models (a.k.a Smart Search)**:
Searching for products and product models at the same time is a little more tricky. 
The rule of thumb is :

```
Given attribute_X and attribute_Y on which a user wishes to search for,
AND
Given an ES indexed list of related product variants and product models (belonging to the same tree).
THEN
Show the node(s) belonging to the level MIN(level(attribute_X), level(attribute_Y)) within that tree.
```

You can find more example on @Delphineray 's search dataset and search use cases.

Example of a tree varying on the color:
```
{
  "identifier": "Root_product",
  "material":  "polyester",
  "product_type": "RootProductModel",
  "owned_attributes": ["material"]
}
{
  "identifier": "product_variant_red"
  "material": "polyester" <--- inherited property
  "color": "red",
  "product_type": "productVariant",
  "owned_attributes": ["color"]
}
{
  "identifier": "product_variant_blue"
  "material": "polyester" <--- inherited property
  "color": "blue",
  "product_type": "productVariant",
  "owned_attributes": ["color"]
}
```

- Searching `Color IN ["blue", "red"]` will return [`"product_variant_blue"`, `"product_variant_red"`]
- Searching `Material = "Polyester"` will return`"Root_product"`
- Searching for `Material = "Polyester" AND Color = "red"` will return `"product_variant_red"` (and not the `"Root_product"`)

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | -
| Added Behats                      | -
| Added integration tests           | Ok
| Changelog updated                 | -
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | -
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Pim/Bundle/CatalogBundle/tests/integration/Elasticsearch/IndexConfiguration/AbstractPimCatalogTestCase.php [hunk 1]
```diff
@@ -70,6 +68,8 @@ protected function indexProducts(array $products)
     protected function getSearchQueryResults(array $query)
     {
         $identifiers = [];
+
+        $query['size'] = 100;
         $response = $this->esClient->search(self::DOCUMENT_TYPE, $query);
 
         foreach ($response['hits']['hits'] as $hit) {
```
Comment (SamirBoulil) [inline src/Pim/Bundle/CatalogBundle/tests/integration/Elasticsearch/IndexConfiguration/AbstractPimCatalogTestCase.php]: @jjanvierI reverted the assertProduct below (you inversed the two, but I didn't understand why :/)

File: src/Pim/Bundle/CatalogBundle/tests/integration/Elasticsearch/IndexConfiguration/AbstractPimCatalogTestCase.php [hunk 2]
```diff
@@ -70,6 +68,8 @@ protected function indexProducts(array $products)
     protected function getSearchQueryResults(array $query)
     {
         $identifiers = [];
+
+        $query['size'] = 100;
```
Comment (juliensnz) [inline src/Pim/Bundle/CatalogBundle/tests/integration/Elasticsearch/IndexConfiguration/AbstractPimCatalogTestCase.php]: constant?

Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/tests/integration/Elasticsearch/IndexConfiguration/AbstractPimCatalogTestCase.php]: we could use a private constant

File: src/Pim/Bundle/CatalogBundle/tests/integration/Elasticsearch/IndexConfiguration/AbstractPimCatalogTestCase.php (hunk 1 above)
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/tests/integration/Elasticsearch/IndexConfiguration/AbstractPimCatalogTestCase.php]: to use the same order than phpunit?

File: src/Pim/Bundle/CatalogBundle/Resources/elasticsearch/index_configuration.yml [hunk 3]
```diff
@@ -22,6 +22,10 @@ mappings:
             identifier:
                 type: 'keyword'
                 normalizer: 'identifier_normalizer'
+            product_type:
+                type: 'keyword'
+            owned_attributes:
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Resources/elasticsearch/index_configuration.yml]: `own_attributes` ?

File: src/Pim/Bundle/CatalogBundle/tests/integration/Elasticsearch/IndexConfiguration/SearchProductsIntegration.php [hunk 4]
```diff
@@ -0,0 +1,762 @@
+<?php
+
+namespace Pim\Bundle\CatalogBundle\tests\integration\Elasticsearch\IndexConfiguration;
+
+/**
+ * Search use cases of products only (no product models should be returned).
+ *
+ * The search is performed independently from the family variants.
+ *
+ * The search takes advantage of the following properties to elaborate concise but powerful requests that retrieves
```
Comment (grena) [inline src/Pim/Bundle/CatalogBundle/tests/integration/Elasticsearch/IndexConfiguration/SearchProductsIntegration.php]: _retrieve_

File: src/Pim/Bundle/CatalogBundle/tests/integration/Elasticsearch/IndexConfiguration/SearchProductsIntegration.php [hunk 5]
```diff
@@ -0,0 +1,762 @@
...
+ * only products and product variants.
+ *
+ * - Each document (e.g: products, product variants or models) has all the properties of it's associated parent model
```
Comment (grena) [inline src/Pim/Bundle/CatalogBundle/tests/integration/Elasticsearch/IndexConfiguration/SearchProductsIntegration.php]: _its_

Comment (SamirBoulil) [inline src/Pim/Bundle/CatalogBundle/tests/integration/Elasticsearch/IndexConfiguration/SearchProductsIntegration.php]: done

File: src/Pim/Bundle/CatalogBundle/tests/integration/Elasticsearch/IndexConfiguration/SearchProductsIntegration.php (hunk 4 above)
Comment (SamirBoulil) [inline src/Pim/Bundle/CatalogBundle/tests/integration/Elasticsearch/IndexConfiguration/SearchProductsIntegration.php]: done

File: src/Pim/Bundle/CatalogBundle/tests/integration/Elasticsearch/IndexConfiguration/SearchProductsAndModelsIntegration.php [hunk 6]
```diff
@@ -0,0 +1,593 @@
...
+
+/**
+ * Search use cases of products and models in a "smart datagrid way".
+ * It returns either products or models depending on where is information is stored.
+ *
+ * The search is performed independently from the family variants.
+ *
+ * The search takes advantage of the following properties to elaborate concise but powerful requests:
+ *
+ * - Each document (e.g: products, product variants or models) has all the properties of it's associated parent model
```
Comment (SamirBoulil) [inline src/Pim/Bundle/CatalogBundle/tests/integration/Elasticsearch/IndexConfiguration/SearchProductsAndModelsIntegration.php]: done

File: src/Pim/Bundle/CatalogBundle/tests/integration/Elasticsearch/IndexConfiguration/SearchProductsAndModelsIntegration.php [hunk 7]
```diff
@@ -0,0 +1,593 @@
+<?php
+
+namespace Pim\Bundle\CatalogBundle\tests\integration\Elasticsearch\IndexConfiguration;
+
+/**
+ * Search use cases of products and models in a "smart datagrid way".
+ * It returns either products or models depending on where is information is stored.
```
Comment (SamirBoulil) [inline src/Pim/Bundle/CatalogBundle/tests/integration/Elasticsearch/IndexConfiguration/SearchProductsAndModelsIntegration.php]: done

File: src/Pim/Bundle/CatalogBundle/tests/integration/Elasticsearch/IndexConfiguration/AbstractPimCatalogTestCase.php (hunk 2 above)
Comment (SamirBoulil) [inline src/Pim/Bundle/CatalogBundle/tests/integration/Elasticsearch/IndexConfiguration/AbstractPimCatalogTestCase.php]: done

**Pull request merged**